components/server/ee/src/prebuilds/prebuild-manager.ts

```diff
@@ -388,17 +388,39 @@ export class PrebuildManager {
         }
     }
 
-    private async shouldRateLimitPrebuild(span: opentracing.Span, cloneURL: string): Promise<boolean> {
+    private async shouldRateLimitPrebuild(span: opentracing.Span, cloneURL: string, userId: string): Promise<boolean> {
         const windowStart = secondsBefore(new Date().toISOString(), PREBUILD_LIMITER_WINDOW_SECONDS);
+
+        // per cloneULR rate limit
         const unabortedCount = await this.workspaceDB
             .trace({ span })
-            .countUnabortedPrebuildsSince(cloneURL, new Date(windowStart));
+            .countUnabortedPrebuildsPerCloneURLSince(cloneURL, new Date(windowStart));
         const limit = this.getPrebuildRateLimitForCloneURL(cloneURL);
 
         if (unabortedCount >= limit) {
-            log.debug("Prebuild exceeds rate limit", { limit, unabortedPrebuildsCount: unabortedCount, cloneURL });
+            log.debug({ userId }, "Prebuild exceeds per-cloneURL rate limit", {
+                limit,
+                unabortedPrebuildsCount: unabortedCount,
+                cloneURL,
+            });
             return true;
         }
+
+        // per user rate limit
```

Comment:
It would've prevented [this](https://gitpod.slack.com/archives/C04BGTDQCSH/p1668778332100729). The traffic did not originate from a single cloneURL

Reply:
@easyCZ, those events did rush in for a single repository, which is in fact a single `cloneURL`. There were different context URL, but this is not relevant. 

To extend on that, the intention of this change is clear and it perfectly makes sense. The issue I want to point out is, that in a case where such a load of events is rushing in, checking against the DB only is less effective. Too many of those checks will be answered equally as they are processes in parallel. 

A holistic solution to that would be to run a control loop on the events recorded and process them in a queue. That way, the DB check would actually be effective.

Reply:
> A holistic solution to that would be to run a control loop on the events recorded and process them in a queue. That way, the DB check would actually be effective.

Absolutely agree. But for now I feel this is reasonable cheap, and a gap in the sense that we are not properly protected against malicious actors.

Regarding the incident: Yes, I got trapped by the cloneURL/contextURL thing at first. But in the end decided that a per-user limit is still missing. If the incident highlighted anything than that we need to be better prepared for these kind of events.

BTW: I think the biggest contributor to the reate-limiting failing is the number of pods involved with handling them.

Reply:
The change adds another "prebuilds are bound to users" dependency. AFAIK we all agree that we want to get rid of that.

Reply:
> But in the end decided that a per-user limit is still missing.

That's just true. 

Let's detach this change from the incident, please. I think we agree this isn't effective in such cases: 1.) per-cloneURL filtering is better suited, and 2.) the race on DB state makes it less effective.

Consider a bad actor would use multiple repositories/webhooks and the per-cloneURL wouldn't catch that. Looking at PREBUILD_LIMITER_WINDOW_SECONDS (~60s), that means that multiple streams of events needs to be coordinated over time to bottom out. Given we're talking about a case involving sophisticated scripting, this would be easily to bypass. I think per-user rate limit requires a wider time window. Compare with API rate limiting on GitHub, where requests of all apps eat up a user's global rate limit (1h), but requests to a single app rate limited within 60s.